File: src/main/java/leets/weeth/domain/attendance/domain/service/AttendanceScheduler.java
package leets.weeth.domain.attendance.domain.service;

import jakarta.transaction.Transactional;
import java.util.List;
import leets.weeth.domain.attendance.domain.entity.Attendance;
import leets.weeth.domain.schedule.domain.entity.Meeting;
import leets.weeth.domain.schedule.domain.service.MeetingGetService;
import lombok.RequiredArgsConstructor;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class AttendanceScheduler {

    private final MeetingGetService meetingGetService;
    private final AttendanceGetService attendanceGetService;
    private final AttendanceUpdateService attendanceUpdateService;

    @Transactional
    @Scheduled(cron = "0 0 22 * * THU", zone = "Asia/Seoul")
    public void autoCloseAttendance() {
        List<Meeting> meetings = meetingGetService.findAllOpenMeetingsBeforeNow();

        meetings.forEach(meeting -> {
                    meeting.close();
                    List<Attendance> attendanceList = attendanceGetService.findAllByMeeting(meeting);
                    attendanceUpdateService.close(attendanceList);
                });
    }
}

File: src/main/kotlin/com/weeth/domain/attendance/domain/service/scheduler/AttendanceScheduler.kt
package com.weeth.domain.attendance.domain.service.scheduler

import com.weeth.domain.attendance.domain.repository.AttendanceRepository
import com.weeth.domain.schedule.domain.service.MeetingGetService
import com.weeth.domain.user.domain.entity.enums.Status
import org.springframework.transaction.annotation.Transactional
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Service

@Service
class AttendanceScheduler(
    private val meetingGetService: MeetingGetService,
    private val attendanceRepository: AttendanceRepository,
) {
    @Transactional
    @Scheduled(cron = "0 0 22 * * THU", zone = "Asia/Seoul")
    fun autoCloseAttendance() {
        val meetings = meetingGetService.findAllOpenMeetingsBeforeNow()

        meetings.forEach { meeting ->
            meeting.close()
            val attendanceList = attendanceRepository.findAllByMeetingAndUserStatus(meeting, Status.ACTIVE)
            attendanceList
                .filter { it.isPending }
                .forEach { attendance ->
                    attendance.close()
                    attendance.user.absent()
                }
        }
    }
}
